File: dotnet/samples/GettingStartedWithProcesses/Step03/Processes/FishAndChipsProcess.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Text.Json;
using Microsoft.SemanticKernel;
using Step03.Models;
using Step03.Steps;

namespace Step03.Processes;

/// <summary>
/// Sample process that showcases how to create a process with a fan in/fan out behavior and use of existing processes as steps.<br/>
/// Visual reference of this process can be found in the <see href="https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/GettingStartedWithProcesses/README.md#fish-and-chips-preparation-process" >diagram</see>
/// </summary>
public static class FishAndChipsProcess
{
    public static class ProcessEvents
    {
        public const string PrepareFishAndChips = nameof(PrepareFishAndChips);
        public const string FishAndChipsReady = nameof(FishAndChipsReady);
        public const string FishAndChipsIngredientOutOfStock = nameof(FishAndChipsIngredientOutOfStock);
    }

    public static ProcessBuilder CreateProcess(string processName = "FishAndChipsProcess")
    {
        var processBuilder = new ProcessBuilder(processName);
        var makeFriedFishStep = processBuilder.AddStepFromProcess(FriedFishProcess.CreateProcess());
        var makePotatoFriesStep = processBuilder.AddStepFromProcess(PotatoFriesProcess.CreateProcess());
        var addCondimentsStep = processBuilder.AddStepFromType<AddFishAndChipsCondimentsStep>();
        // An additional step that is the only one that emits an public event in a process can be added to maintain event names unique
        var externalStep = processBuilder.AddStepFromType<ExternalFishAndChipsStep>();

        processBuilder
            .OnInputEvent(ProcessEvents.PrepareFishAndChips)
            .SendEventTo(makeFriedFishStep.WhereInputEventIs(FriedFishProcess.ProcessEvents.PrepareFriedFish))
            .SendEventTo(makePotatoFriesStep.WhereInputEventIs(PotatoFriesProcess.ProcessEvents.PreparePotatoFries));

        processBuilder.ListenFor().AllOf([
            new(FriedFishProcess.ProcessEvents.FriedFishReady, makeFriedFishStep),
            new(PotatoFriesProcess.ProcessEvents.PotatoFriesReady, makePotatoFriesStep),
        ]).SendEventTo(new ProcessStepTargetBuilder(addCondimentsStep, inputMapping: inputEvents =>
        {
            return new() {
                { "fishActions", inputEvents[makeFriedFishStep.GetFullEventId(FriedFishProcess.ProcessEvents.FriedFishReady)] },
                { "potatoActions", inputEvents[makePotatoFriesStep.GetFullEventId(PotatoFriesProcess.ProcessEvents.PotatoFriesReady)] },
            };
        }));

        addCondimentsStep
            .OnEvent(AddFishAndChipsCondimentsStep.StepEvents.CondimentsAdded)
            .SendEventTo(new ProcessFunctionTargetBuilder(externalStep));

        externalStep.OnEvent(ExternalFishAndChipsStep.StepEvents.FishAndChipsReady).EmitAsPublicEvent();

        return processBuilder;
    }

    public static ProcessBuilder CreateProcessWithStatefulSteps(string processName = "FishAndChipsWithStatefulStepsProcess")
    {
        var processBuilder = new ProcessBuilder(processName);
        var makeFriedFishStep = processBuilder.AddStepFromProcess(FriedFishProcess.CreateProcessWithStatefulStepsV1());
        var makePotatoFriesStep = processBuilder.AddStepFromProcess(PotatoFriesProcess.CreateProcessWithStatefulSteps());
        var addCondimentsStep = processBuilder.AddStepFromType<AddFishAndChipsCondimentsStep>();
        // An additional step that is the only one that emits an public event in a process can be added to maintain event names unique
        var externalStep = processBuilder.AddStepFromType<ExternalFishAndChipsStep>();

        processBuilder
            .OnInputEvent(ProcessEvents.PrepareFishAndChips)
            .SendEventTo(makeFriedFishStep.WhereInputEventIs(FriedFishProcess.ProcessEvents.PrepareFriedFish))
            .SendEventTo(makePotatoFriesStep.WhereInputEventIs(PotatoFriesProcess.ProcessEvents.PreparePotatoFries));

        processBuilder.ListenFor().AllOf(
            [
                new(FriedFishProcess.ProcessEvents.FriedFishReady, makeFriedFishStep),
                new(PotatoFriesProcess.ProcessEvents.PotatoFriesReady, makePotatoFriesStep),
            ]).SendEventTo(new ProcessStepTargetBuilder(addCondimentsStep, inputMapping: inputEvents =>
            {
                return new() {
                    { "fishActions", inputEvents[makeFriedFishStep.GetFullEventId(FriedFishProcess.ProcessEvents.FriedFishReady)] },
                    { "potatoActions", inputEvents[makePotatoFriesStep.GetFullEventId(PotatoFriesProcess.ProcessEvents.PotatoFriesReady)] },
                };
            }));

        addCondimentsStep
            .OnEvent(AddFishAndChipsCondimentsStep.StepEvents.CondimentsAdded)
            .SendEventTo(new ProcessFunctionTargetBuilder(externalStep));

        externalStep.OnEvent(ExternalFishAndChipsStep.StepEvents.FishAndChipsReady).EmitAsPublicEvent();

        return processBuilder;
    }

    private sealed class AddFishAndChipsCondimentsStep : KernelProcessStep
    {
        public static class ProcessFunctions
        {
            public const string AddCondiments = nameof(AddCondiments);
        }

        public static new class StepEvents
        {
            public static readonly KernelProcessEventDescriptor<List<string>> CondimentsAdded = new(nameof(CondimentsAdded));
        }

        [KernelFunction(ProcessFunctions.AddCondiments)]
        public async Task AddCondimentsAsync(KernelProcessStepContext context, List<string> fishActions, List<string> potatoActions)
        {
            Console.WriteLine($"ADD_CONDIMENTS: Added condiments to Fish & Chips - Fish: {JsonSerializer.Serialize(fishActions)}, Potatoes: {JsonSerializer.Serialize(potatoActions)}");
            fishActions.AddRange(potatoActions);
            fishActions.Add(FoodIngredients.Condiments.ToFriendlyString());
            await context.EmitEventAsync(StepEvents.CondimentsAdded, fishActions);
        }
    }

    private sealed class ExternalFishAndChipsStep : ExternalStep
    {
        public static new class StepEvents
        {
            public static readonly KernelProcessEventDescriptor<List<string>> FishAndChipsReady = new(nameof(ProcessEvents.FishAndChipsReady));
        }

        public ExternalFishAndChipsStep() : base(ProcessEvents.FishAndChipsReady) { }
    }
}

File: python/samples/getting_started_with_processes/step03/processes/fish_and_chips_process.py
# Copyright (c) Microsoft. All rights reserved.
import json
from enum import Enum

from samples.getting_started_with_processes.step03.processes.fried_fish_process import (
    FriedFishProcess,
)
from samples.getting_started_with_processes.step03.processes.potato_fries_process import (
    PotatoFriesProcess,
)
from samples.getting_started_with_processes.step03.steps.external_step import ExternalStep
from semantic_kernel.functions.kernel_function_decorator import kernel_function
from semantic_kernel.processes import ProcessBuilder
from semantic_kernel.processes.kernel_process import (
    KernelProcessStep,
    KernelProcessStepContext,
)


class AddFishAndChipsCondimentsStep(KernelProcessStep):
    class Functions(Enum):
        AddCondiments = "AddCondiments"

    class OutputEvents(Enum):
        CondimentsAdded = "CondimentsAdded"

    @kernel_function(name=Functions.AddCondiments)
    async def add_condiments(
        self, context: KernelProcessStepContext, fish_actions: list[str], potato_actions: list[str]
    ):
        print(
            f"ADD_CONDIMENTS: Added condiments to Fish & Chips - "
            f"Fish: {json.dumps(fish_actions)}, Potatoes: {json.dumps(potato_actions)}"
        )
        fish_actions.extend(potato_actions)
        fish_actions.append("Condiments")
        await context.emit_event(
            process_event=self.OutputEvents.CondimentsAdded,
            data=fish_actions,
        )


class FishAndChipsProcess:
    class ProcessEvents(Enum):
        PrepareFishAndChips = "PrepareFishAndChips"
        FishAndChipsReady = "FishAndChipsReady"
        FishAndChipsIngredientOutOfStock = "FishAndChipsIngredientOutOfStock"

    class ExternalFishAndChipsStep(ExternalStep):
        def __init__(self) -> None:
            super().__init__(FishAndChipsProcess.ProcessEvents.FishAndChipsReady)

    @staticmethod
    def create_process(process_name: str = "FishAndChipsProcess") -> ProcessBuilder:
        process_builder = ProcessBuilder(process_name)

        make_fish_step = process_builder.add_step_from_process(FriedFishProcess.create_process())
        make_potato_step = process_builder.add_step_from_process(PotatoFriesProcess.create_process())
        add_condiments_step = process_builder.add_step(AddFishAndChipsCondimentsStep)
        external_step = process_builder.add_step(FishAndChipsProcess.ExternalFishAndChipsStep)

        process_builder.on_input_event(FishAndChipsProcess.ProcessEvents.PrepareFishAndChips).send_event_to(
            make_fish_step.where_input_event_is(FriedFishProcess.ProcessEvents.PrepareFriedFish)
        ).send_event_to(make_potato_step.where_input_event_is(PotatoFriesProcess.ProcessEvents.PreparePotatoFries))

        make_fish_step.on_event(FriedFishProcess.ProcessEvents.FriedFishReady).send_event_to(
            add_condiments_step, parameter_name="fishActions"
        )
        make_potato_step.on_event(PotatoFriesProcess.ProcessEvents.PotatoFriesReady).send_event_to(
            add_condiments_step, parameter_name="potatoActions"
        )

        add_condiments_step.on_event(AddFishAndChipsCondimentsStep.OutputEvents.CondimentsAdded).send_event_to(
            external_step
        )

        return process_builder

    @staticmethod
    def create_process_with_stateful_steps(
        process_name: str = "FishAndChipsWithStatefulStepsProcess",
    ) -> ProcessBuilder:
        process_builder = ProcessBuilder(process_name)

        make_fish_step = process_builder.add_step_from_process(FriedFishProcess.create_process_with_stateful_steps_v1())
        make_potato_step = process_builder.add_step_from_process(
            PotatoFriesProcess.create_process_with_stateful_steps()
        )
        add_condiments_step = process_builder.add_step(AddFishAndChipsCondimentsStep)
        external_step = process_builder.add_step(FishAndChipsProcess.ExternalFishAndChipsStep)

        process_builder.on_input_event(FishAndChipsProcess.ProcessEvents.PrepareFishAndChips).send_event_to(
            make_fish_step.where_input_event_is(FriedFishProcess.ProcessEvents.PrepareFriedFish)
        ).send_event_to(make_potato_step.where_input_event_is(PotatoFriesProcess.ProcessEvents.PreparePotatoFries))

        make_fish_step.on_event(FriedFishProcess.ProcessEvents.FriedFishReady).send_event_to(
            add_condiments_step, parameter_name="fishActions"
        )
        make_potato_step.on_event(PotatoFriesProcess.ProcessEvents.PotatoFriesReady).send_event_to(
            add_condiments_step, parameter_name="potatoActions"
        )

        add_condiments_step.on_event(AddFishAndChipsCondimentsStep.OutputEvents.CondimentsAdded).send_event_to(
            external_step
        )

        return process_builder
